### PR TITLE
feat: add route selection panel

### PR DESCRIPTION
**Pull request by firestack**
Depends On:
- #2645
- #2648 
- #2649 
- #2650 
- #2651 

Asana Ticket: https://app.asana.com/0/1203014709808707/1207372095414476/f

### CODE REVIEW

File: assets/css/bootstrap/_map_overrides.scss
Comment (joshlarson) [inline assets/css/bootstrap/_map_overrides.scss:16]: I'm curious about this comment - are `top-0`, etc examples of possible utility classes?

Comment (joshlarson) [reply assets/css/bootstrap/_map_overrides.scss:16]: I ask because unless I'm missing something special about the `$utilities` syntax, we're only actually short-handing `inset`?

File: assets/src/models/createDetourMachine.ts
Comment (joshlarson) [inline assets/src/models/createDetourMachine.ts:29]:
```suggestion
```

I don't think this is a use case that we currently want to support. The two entrypoints are that we already have a route pattern (from the search map), or that we have a route but no route pattern ( from the route ladders).

File: assets/src/models/createDetourMachine.ts
Comment (joshlarson) [inline assets/src/models/createDetourMachine.ts:90]:
```suggestion
          initial: "Load Route Patterns",
```

Related to https://github.com/mbta/skate/pull/2652/files#r1630163656, I think we'll land in this state when the input has a route, but no route pattern, and when that happens, we want to be loading the route patterns.

(I had to make that change [here](https://github.com/mbta/skate/pull/2647/files#diff-2697d2574fc8e2dec397eba594a8b6fa60a8f51bf3ea7fa188e87d697a87bb50R90) in order to get the detour modal to load properly, but I think it makes more sense to make that change in this PR.

File: assets/src/models/createDetourMachine.ts
Comment (firestack) [reply assets/src/models/createDetourMachine.ts:29]: This is not something we _currently_ have implemented, but it would not be difficult to update the `<DummyDetourPage>` to use this behavior.

I think we _do_ want this behavior, as per this [figma slide: Detour Opened from Detours Page](https://www.figma.com/design/XGmmwVROXdOvnWtJeUFC50/Skate-%7C-Unplanned-Detours-%7C-Hand-Off?node-id=2855-6968&t=XF9NKdbU2WB3iojO-1), but you could argue it shouldn't be implemented yet because that's not what we're doing yet. 

IMO, I'd leave it so that it's clear in the future what the machine _can_ handle, but we could drop it.

File: assets/src/models/createDetourMachine.ts
Comment (joshlarson) [reply assets/src/models/createDetourMachine.ts:90]: UPDATE: ☝️ was based on an old version of this branch, and is no longer true. 

In fact, making my change would break this because the `Load Route Patterns` state doesn't exist anymore!

File: assets/src/components/detours/detourRouteSelectionPanel.tsx
Comment (joshlarson) [inline assets/src/components/detours/detourRouteSelectionPanel.tsx:103]: Not exactly a nit, but a UX nit. This works GREAT if we're transitioning from a state with route patterns to another state with route patterns,

https://github.com/mbta/skate/assets/912020/107a02b5-5383-4f1f-8e95-ed35f22bbc73

...but is a much less smooth UX when you're starting with no route patterns

https://github.com/mbta/skate/assets/912020/958155c5-8d5f-4d36-8a0d-822b05c2d905

This can be a follow-up ticket, but I just wanted to flag it.

File: assets/src/models/createDetourMachine.ts
Comment (firestack) [reply assets/src/models/createDetourMachine.ts:105]: > The current `FormSelect` implementation has an entry with a `null` route ID also (I think), which should also trigger this event.

Yes, and just for clarity, I've explicitly handled that state instead of ignoring it https://github.com/mbta/skate/pull/2652/files#diff-a5c2ebb8d72dc32209d3598dadecd6793515e43155078116e8176cc292c32a0aR162-R173


---

I'll go ahead and edit this comment to remove the project requirement bit and probably the test bit?

Comment (firestack) [reply assets/src/models/createDetourMachine.ts:105]: I'm deciding to remove the comment altogether, but if you disagree, lets chat about what to put instead.

File: assets/src/components/detours/diversionPage.tsx
Comment (joshlarson) [inline assets/src/components/detours/diversionPage.tsx:145]: Non-blocking: Would it make sense to lift the `Pick Route Pattern` and `Share Detour` machines to top level? I would expect the top-level states to basically line up with the panels on the left, and it surprised me a bit to see `Detour Drawing` as part of the condition that governs the panel that shows up _before_ you start drawing.

Comment (joshlarson) [reply assets/src/components/detours/diversionPage.tsx:145]: Non-blocking for a few reasons, one of which is that I don't know enough about state machines / xstate to know whether this is even a good idea.

But also because I think it would be good for me or Hannah to try our hands at doing a state machine refactor like this.

And most importantly because I don't want to block merging of this (and the work that depends on this) on another style change.

Comment (firestack) [reply assets/src/components/detours/diversionPage.tsx:145]: > But also because I think it would be good for me or Hannah to try our hands at doing a state machine refactor like this.

I agree with this! Maybe open a ticket and put it on the [Skate Tech Cleanup board](https://app.asana.com/0/1152340551558956/1204993290118919)?



> Non-blocking: Would it make sense to lift the `Pick Route Pattern` and `Share Detour` machines to top level? I would expect the top-level states to basically line up with the panels on the left, and it surprised me a bit to see `Detour Drawing` as part of the condition that governs the panel that shows up _before_ you start drawing.

I'm not really sure on this, I initially nested them in case we needed say, parallel states running at all times, or other modes we haven't thought of. Since there's no real easy LSP way of refactoring the `matches` calls, I wouldn't opt to do this _myself_ but I don't mind entertaining the idea and trying it if one of you would like to make that refactor.

File: assets/tests/components/detours/diversionPage.test.tsx
Comment (joshlarson) [inline assets/tests/components/detours/diversionPage.test.tsx:1236]: I think the comment above this can go away now.

File: assets/tests/components/detours/diversionPage.test.tsx
Comment (joshlarson) [inline assets/tests/components/detours/diversionPage.test.tsx:1549]: Nit: Should this be `const rpcRadioButtonName = (rp2: RoutePattern) => {...}` instead (and probably live at the top of the file 😩)?

Other nit: The variable name should perhaps just be `rp`?

---

Also...

I also would sooner see the tests be a bit more explicit about their data, so something like

```typescript
rp = routePatternFactory.build({
  name: "Harvard via Allston",
  description: "from Nubian Station",
})

/* ... */

screen.getByRole("radio", {
  name: "Harvard via Allston from Nubian Station",
})
```

That way the test is super explicit about what it's testing (and it's clearer what type of thing we expect `name` and `description` to be). As it currently stands, in order to figure out what the assertion actually is, we have to trace through `RPCRadioButtonName`, and if `patternDisplayName` is broken then the assertion will assert the broken/wrong thing.

This can be a follow-up, and doesn't need to block merging this PR.

**Review by joshlarson — CHANGES_REQUESTED** [review]
Arg - I didn't mean to make my comment a review comment. 

The first part (the two nits) I think are worth changing before merging. The other part is good to be a follow-up. 

I'm close to done, but not quite done, reviewing this!

File: assets/tests/components/detours/diversionPage.test.tsx
Comment (firestack) [reply assets/tests/components/detours/diversionPage.test.tsx:1604]: Does it make it any better that `patternDisplayName` is tested? https://github.com/mbta/skate/blob/main/assets/tests/components/mapPage/routePropertiesCard.test.tsx#L58-L104

File: assets/tests/components/detours/diversionPage.test.tsx
Comment (firestack) [reply assets/tests/components/detours/diversionPage.test.tsx:1231]: https://github.com/mbta/skate/compare/67e118cb595b85d64dd29f50cd987ca1a0776045..a0147d026bc04ce11131c677e0e07c2f18d9bc0b

File: assets/tests/components/detours/diversionPage.test.tsx
Comment (joshlarson) [inline assets/tests/components/detours/diversionPage.test.tsx:1284]: **Non-blocking**

In general, I don't love having a `beforeEach` in the middle of the test file, because it makes it a lot harder to tell what context each test is working with. For any of the tests within this `describe`, they've got an extra test group enabled, but it's hard to tell when looking at a given test which `describe` it's in.

A possible solution would be to enable this test group at the top for _all_ tests in this file, and then disable it only for the one test that asserts that the button isn't there (I think there's only one test that cares).

Non-blocking because this will be short-lived anyway, but food for thought.

File: assets/tests/components/detours/diversionPage.test.tsx
Comment (firestack) [reply assets/tests/components/detours/diversionPage.test.tsx]: Part of it https://github.com/mbta/skate/compare/356f1eb5c81370545b67bc2e28304b7595914805..aeb951f1a4020482c2a6040753fe5bfc440d5de2

File: assets/tests/components/detours/diversionPage.test.tsx
Comment (joshlarson) [inline assets/tests/components/detours/diversionPage.test.tsx:70]:
```suggestion
  return name + " " + description
```

Nit

File: assets/src/models/createDetourMachine.ts
Comment (joshlarson) [reply assets/src/models/createDetourMachine.ts:29]: Eh that's fair. Since we're pretty certain we'll want to support it, I'm okay leaving it in.

**Review by joshlarson — APPROVED** [review]
Looks great!

I raised a few things as follow-ups, but those (discussions and/or fixes) can happen later!

File: assets/tests/components/detours/diversionPage.test.tsx
Comment (joshlarson) [inline assets/tests/components/detours/diversionPage.test.tsx:1502]:
```suggestion
    test("when route is changed from a route to empty, can still select another route", async () => {
```

Tiny grammar nit.

**Review by joshlarson — APPROVED** [review]

**Pull request merged**
